src/scripts/components/App/App.tsx

```diff
@@ -0,0 +1,151 @@
+import { Component } from 'react';
+import Search from '../Search/Search';
+import Cards from '../Cards/Cards';
+import HpApi from '../../api/HpApi';
+import { PotionsReqParams, PotionsResponse } from '../../api/types/potions';
+import Pagination from '../Pagination/Pagination';
+import defineNumberOfPages from '../Pagination/defineNumberOfPages';
+import BtnError from '../BtnError/BtnError';
+import Loader from '../Preloader/Preloader';
+
+const categories = [
+  'characteristics',
+  'difficulty',
+  'effect',
+  'inventors',
+  'ingredients',
+  'manufacturers',
+  'name',
+  'side_effects',
+  'time',
+];
+
+interface IState {
+  isLoaded: boolean;
+  items: PotionsResponse['data'] | null;
+  pagination: {
+    current: number;
+    last: number;
+    next: number;
+    pages: number;
+  };
+  params: PotionsReqParams;
+}
+
+const defaultPotionParams = {
+  sort: { param: 'ASC', attribute: 'name' },
+  filters: undefined,
+  pagination: { limit: 30, page: 1 },
+};
+const lsPotionParams = localStorage.getItem('potionsParams');
+export default class App extends Component<unknown, IState> {
+  constructor(props: unknown) {
+    super(props);
+    this.state = {
+      isLoaded: false,
+      items: null,
+      pagination: {
+        current: 1,
+        last: 0,
+        next: 0,
+        pages: 0,
+      },
+      params: lsPotionParams ? JSON.parse(lsPotionParams) : defaultPotionParams,
+    };
+  }
+
+  public handlePaginationClick = (btn: 'start' | 'next' | 'prev' | 'end') => {
+    let page = this.state.params.pagination.page;
+    switch (btn) {
+      case 'start':
+        page = 1;
+        break;
+      case 'prev':
+        page = page - 1;
+        break;
+      case 'next':
+        page = page + 1;
+        break;
+      case 'end':
+        page = this.state.pagination.pages;
+        break;
+      default:
+        break;
+    }
+    const newParams = { ...this.state.params };
+    newParams.pagination.page = page;
+    this.setState({ params: newParams });
+  };
+
+  public handleSendParams = async (filter: PotionsReqParams['filters']) => {
+    this.setState((state) => ({
+      params: {
+        filters: filter,
+        pagination: {
+          page: 1,
+          limit: state.params.pagination.limit,
+        },
+      },
+    }));
+  };
+
+  public async getData() {
+    this.setState({ isLoaded: false });
+    const res = await HpApi.getPotions(this.state.params);
+    if (!res) return;
+    this.setState((state) => ({
+      items: res.data,
+      pagination: {
+        current: res.meta.pagination.current,
+        last: res.meta.pagination.last ? res.meta.pagination.last : 0,
```

Comment:
You need to make sure that there are `meta.pagination.<property>`. In case it's not there there will be an error. You can use 2 options:
1. make sure in if condition that there is such object
2. you can use optional chaining 

Also you can combine them